api/src/certification/complementary-certification/domain/usecases/attach-badges-to-complementary-certification.js

```diff
@@ -0,0 +1,154 @@
+import { NotFoundError } from '../../../../../lib/domain/errors.js';
+import { DomainTransaction } from '../../../../../lib/infrastructure/DomainTransaction.js';
+import uniq from 'lodash/uniq.js';
+
+import bluebird from 'bluebird';
+import { ComplementaryCertificationBadgeToAttach } from '../models/ComplementaryCertificationBadgeToAttach.js';
+import { InvalidBadgeLevelError } from '../../../shared/domain/errors.js';
+
+const attachBadgesToComplementaryCertification = async function ({
+  complementaryCertificationId,
+  userId,
+  targetProfileIdToDetach,
+  complementaryCertificationBadgesToAttachDTO,
+  complementaryCertificationBadgeToAttachValidator,
+  badgeRepository,
+  complementaryCertificationRepository,
+  complementaryCertificationBadgesRepository,
+}) {
+  _verifyThatLevelsAreConsistent({
+    complementaryCertificationBadgesToAttachDTO,
+  });
+
+  await _verifyThatComplementaryCertificationExists({
+    complementaryCertificationId,
+    complementaryCertificationRepository,
+  });
+
+  await _verifyThatBadgesToAttachExist({
+    complementaryCertificationBadgesToAttachDTO,
+    badgeRepository,
+  });
+
+  const complementaryCertificationBadgesToAttach = complementaryCertificationBadgesToAttachDTO.map(
+    (complementaryCertificationBadgeToAttachDTO) => {
+      const complementaryCertificationBadgeToAttach = ComplementaryCertificationBadgeToAttach.from({
+        ...complementaryCertificationBadgeToAttachDTO,
+        complementaryCertificationId,
+        userId,
+      });
+      complementaryCertificationBadgeToAttachValidator.validate(complementaryCertificationBadgeToAttach);
+      return complementaryCertificationBadgeToAttach;
+    },
+  );
+
+  return DomainTransaction.execute(async (domainTransaction) => {
+    const relatedComplementaryCertificationBadgesIds =
+      await complementaryCertificationBadgesRepository.getAllIdsByTargetProfileId({
+        targetProfileId: targetProfileIdToDetach,
+      });
+
+    if (relatedComplementaryCertificationBadgesIds.length === 0) {
+      throw new NotFoundError('No badges for this target profile.');
+    }
+
+    await _detachExistingComplementaryCertificationBadge({
+      complementaryCertificationBadgesRepository,
+      relatedComplementaryCertificationBadgesIds,
+      domainTransaction,
+    });
+
+    await bluebird.mapSeries(
+      complementaryCertificationBadgesToAttach,
+      async (complementaryCertificationBadgeToAttach) => {
+        await _attachNewComplementaryCertificationBadge({
+          complementaryCertificationBadgesRepository,
+          complementaryCertificationBadge: complementaryCertificationBadgeToAttach,
+          domainTransaction,
+        });
+      },
+    );
```

Comment:
L'usage de bluebird dans le cadre du `mapSeries` est superflu ici. Nous ne récupérons rien du `map` ce qui fait que nous n'avons pas besoin d'une version 'async' du Array.map.
La séquentialité peut être respectée avec un for..of ce qui nous évite d'utiliser du coup une lib pour faire du natif

```suggestion
    for (const complementaryCertificationBadge of complementaryCertificationBadgesToAttach) {
      await _attachNewComplementaryCertificationBadge({
        complementaryCertificationBadgesRepository,
        complementaryCertificationBadge,
        domainTransaction,
      });
    }
```